# reGenotypeTE
This pipeline is currenly under construction by Jainy Thomas (University of Utah) and Clement Goubert (Cornell University).
Elaborated with the collaboration of Jeffrey M. Kidd (University of Michigan)

## Purpose

reGenotypeTE is a pipeline dedicated to re-genotype Mobile Element Insertion (MEI) previously scored with MELT (Mobile Element Locator Tool, ). reGenotypeTE extracts reads from previously detected polymorphic MEI and performs using populations data a local re-assembly of both presence and absence loci. Eventually, remapping of the reads at the infividual level allow to score the genotype of the MEI using a modified version of Li's et al. genotype likelihood. This method drammatically improves the quality of the genotypes of reported MEI and can be directly used after a MELT run on both de-novo (non-reference) and deletion (reference) insertions.

## Installation

### Dependencies

These programs need to be installed and their path reported in the file "parameterfile.init"
Perl executable must be in the user path

* PERL https://www.perl.org/
* PARALLEL https://www.gnu.org/software/parallel/
* PICARD https://broadinstitute.github.io/picard/
* BEDTOOLS http://bedtools.readthedocs.io/en/latest/
* SEQTK https://github.com/lh3/seqtk
* BAMUTILS https://genome.sph.umich.edu/wiki/BamUtil
* SPADES http://cab.spbu.ru/software/spades/
* MINIA http://minia.genouest.org/
* CAP3 http://seq.cs.iastate.edu/cap3.html
* BLAST ftp://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/
* BWA http://bio-bwa.sourceforge.net/bwa.shtml
* BGZIP http://www.htslib.org/doc/bgzip.html
* TABIX http://www.htslib.org/doc/tabix.html

### Download and install

Clone from git repository:

```sh
git clone https://github.com/clemgoub/reGenotypeTE
cd reGenotypeTE
```

## De-novo insertion (non-reference)

### File preparation
1. a vcf/vcf.gz file generated by the MELT discovery workflow. If the loci/individuals are sampled from the original vcf/vcf.gz use the following flag `--recode-INFO-all` in vcftools so the subsetted vcf will be compatible with reGenotypeTE

2. bam files for each individual found in the vcf file

3. a two column tab separated table with the sample name and corresponding bam name:

```
sample1 sample1-xxx-file.bam
sample2 sample2-yyy-file.bam
sample3 sample3-zzz-file.bam
```
4. Reference genome in fasta format (to date tested with hg19 and hg38)

5. Edit the file "parameterfile.init" following the indications within:

```sh
### MAIN PARAMETERS
## input
# user data
VCF="/home/xxx/MELT.vcf" #Path to MELT vcf (.vcf or .vcf.gz) must contain INFO field with TSD and MEI type (1)
BAMPATH="/vbod2/xxx/bams" # Path to the bams folder (2)
BAMFILE="/home/xxx/bams.txt" # <indiv_name> <bam_name> (two fields tab separated table) (3)
# genome data
RM_TRACK="./Ressources/RepeatMasker_Alu_hg38_lift_hg19.bed" # set by default for hg19 / use ./Ressources/RepeatMasker_Alu_hg38.bed
RM_FASTA="./Ressources/refinelib" # set by default to be compatible with the Repeat Masker track included in the package
# output
OUTDIR="/home/cgoubert/CorrectHet" # Path to place the output directory (will be named after PROJECT); OUTDIR must exist
PROJECT="TEST_400-100" # Name of the project (name of the folder)

### OPTIONS
## mendatory parameters
individual_nb="10" # number of individual per job (try to minimize that number)
CPU="40" # number of CPU (try to maximize that number such as CPU x individual_nb = total nb of individuals)
GENOME="" # Path the the reference genome sequence
## non-mendatory parameters
MAP="NO" #OR NO # to give mappability score of MEI (experimental)

### DEPENDENCIES PATH
PARALLEL="/usr/bin/parallel" #Path to the GNU Parallel program (executable)
PICARD="/home/xxx/software/picard-2.9.2" #Path to Picard Tools (executable)
BEDTOOLS="/home/xxx/bin/bedtools2/bin" #Path to bedtools (folder)
SEQTK="/home/xxx/software/seqtk" (folder)
BAMUTILS="/home/xxx/software/bamUtil" (folder)
SPADES="/home/xxx/software/SPAdes-3.11.1-Linux/bin" #Path to spades bin directory (to locate spades.py and dispades.py)
MINIA="/home/xxx/software/minia-v2.0.7-Source/build/bin" #Path to minia bin directory
CAP3="/home/xxx/software/CAP3" #Path to CAP3 directory
BLAST="/home/xxx/software/ncbi-blast-2.6.0+/bin" #Path to blast bin directory
BWA="/home/xxx/bin/bwa-0.7.16a/bwa" #Path to bwa
BGZIP="/usr/local/bin/bgzip" #Path to bgzip
TABIX="/usr/local/bin/tabix" #Path to tabix
# /!\ PERL MUST BE IN PATH /!\
```

### Run reGenotypeTE

Once your files are ready and parameterfile.init properly filled, run the following command in the reGenotypeTE folder:

```sh
nohup ./run_reGenotypeTE.sh &> reGenotypeTE.log &
```

#### Test run
We have prepared a small tutorial/test-run to check if all the components of reGenotypeTE works perfectly.

We are going to run the pipeline on 2 loci of 3 individuals from the 1000 Genome Project.

1. Download the bam and bam.bai files

Within the reGenotypeTE folder, type:
```sh
cd test_data
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase3/data/NA07056/alignment/NA07056.mapped.ILLUMINA.bwa.CEU.low_coverage.20130415.bam
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase3/data/NA07056/alignment/NA07056.mapped.ILLUMINA.bwa.CEU.low_coverage.20130415.bam.bai
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase3/data/NA11830/alignment/NA11830.mapped.ILLUMINA.bwa.CEU.low_coverage.20120522.bam
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase3/data/NA11830/alignment/NA11830.mapped.ILLUMINA.bwa.CEU.low_coverage.20120522.bam.bai
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase3/data/NA12144/alignment/NA12144.mapped.ILLUMINA.bwa.CEU.low_coverage.20130415.bam
wget ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase3/data/NA12144/alignment/NA12144.mapped.ILLUMINA.bwa.CEU.low_coverage.20130415.bam.bai
```
The corresponding bam/bam.bai files will be downladed into <yourpath>/reGenotypeTE/test_data
  
2. Copy the parameterfile.init present in <yourpath>/reGenotypeTE/test_data to the main folder
  
```sh
cp parameterfile.init ../
cd ../
```

3. Edit the parameterfile.init according to your dependancies and local path (but do not change anything else!)

4. Run reGenotypeTE

```sh
nohup ./run_reGenotypeTE.sh &> reGenotypeTE_TESTRUN.log &
```
5. Expected results

The genotype from the original vcf (<>/reGenotypeTE/test_data/testDiscoveryAlu.vcf) are the following

|         | 1_72639020 | 10_69994906 |
|---------|------------|-------------|
| NA07056 | 1          | 0           |
| NA11830 | 0          | 1           |
| NA12155 | 1          | 1           |

After running the test, you can convert your output vcf.gz into a 012 table using:

```
vcftools --gzvcf TEST_dataALU.reGenotypeTE.vcf.gz --012
```

The new genotype should be 

|         | 1_72639020 | 10_69994906 |
|---------|------------|-------------|
| NA07056 | 2          | 0           |
| NA11830 | 0          | 2           |
| NA12155 | 2          | 1           |


## Deletions (reference-insertions)
Soon! We are currently testing it!
